src/data/matrix/basic.lean

```diff
@@ -26,24 +26,120 @@ open dmatrix
+instance [semiring R] [add_comm_monoid α] [module R α] :
+  module R (matrix m n α) := pi.module _ _ _
+
 section ext
+
 variables {M N : matrix m n α}
 
 theorem ext_iff : (∀ i j, M i j = N i j) ↔ M = N :=
 ⟨λ h, funext $ λ i, funext $ h i, λ h, by simp [h]⟩
 
-@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N :=
-ext_iff.mp
+@[ext] theorem ext : (∀ i j, M i j = N i j) → M = N := ext_iff.mp
 
 end ext
 
+section transpose
+
+/-- The transpose of a matrix. -/
+def transpose (M : matrix m n α) : matrix n m α
+| x y := M y x
+localized "postfix `ᵀ`:1500 := matrix.transpose" in matrix
+
+lemma trans_col_eq_row (A : matrix m n α) (i : m) : (λ j, Aᵀ j i) = A i :=
+by simp [transpose]
+lemma trans_row_eq_col (A : matrix m n α) (j : n) : Aᵀ j = (λ i, A i j):=
+by ext; simp [transpose]
+
+instance vec_has_star [has_star α]: has_star (n → α) := ⟨λ v i, star (v i)⟩
+@[simp] lemma vec_star_ext [has_star α] (v : n → α) (i : n) : star v i = star (v i) := rfl
+
+def conj_transpose [has_star α] (M : matrix m n α) : matrix n m α
+| x y := star (M y x)
+localized "postfix `ᴴ`:1500 := matrix.conj_transpose" in matrix
+
+end transpose
+
+open_locale matrix
+
+protected def is_sym (A : matrix m m α) : Prop := Aᵀ = A
+
+protected def is_skewsym [has_neg α] (A : matrix m m α) : Prop := -Aᵀ = A
+
+protected def is_Hermitian [has_star α] (A : matrix m m α) : Prop := Aᴴ = A
+
+/-- Proposition `matrix.is_perm`.
+    `P.is_perm` means `P` is a permutation matrix.
+    A new file is probably required for systematically investigating permutation matrices. -/
+protected def is_perm [decidable_eq m] [has_zero α] [has_one α] (P : matrix m m α): Prop :=
+∃ σ : equiv.perm m, P = equiv.perm.to_matrix σ
+
+/-- Proposition `matrix.is_perfect_shuffle`.
+    `P.is_perfect_shuffle` means `P` is a perfect shuffle matrix.
+    A new file is probably required for systematically investigating perfect shuffle matrices. -/
+protected def is_perfect_shuffle
+[decidable_eq m] [has_zero α] [has_one α] (P : matrix m m α) : Prop :=
+∃ σ : equiv.perm m, (P = equiv.perm.to_matrix σ ∧ ∀ i : m, σ i ≠ i)
+
+/-- The "trace" of a matrix.
+    A different version of "trace" is defined in `trace.lean` as `matrix.trace`.
+    One advantage of `matrix.tr` is that this definition is more elementary,
+    which only requires α to be a `add_comm_monoid`; whilst `matrix.trace` requires
+    `[semiring β] [add_comm_monoid α] [module β α]`.
+    The equivalence can be easily established when `α` is indeed a `β-module`.
+    Another advantage is that `matrix.tr` is more convenient for users to explore lemmas/theorems
+    involving "trace" from a combinatorial aspect.-/
+def tr [add_comm_monoid α] (A : matrix n n α) : α := ∑ i : n, A i i
+/-
+lemma trace_eq_tr [semiring β] [add_comm_monoid α] [module β α] (A : matrix n n α)
+: trace n β α A = tr A := rfl
+-/
+
+lemma eq_of_empty [c: is_empty m] (M N: matrix m m α) : M = N :=
+by {ext, exfalso, apply is_empty_iff.mp c i}
```

Comment:
This lemma already exists, it's `subsingleton.elim M N` after doing `local attribute [instance] subsingleton_of_empty_left subsingleton_of_empty_right`.

Reply:
`matrix.subsingleton_of_empty_left` and the other report undefined.

Reply:
All other changes are now done.

Reply:
It's showing as undefined because you're trying to use it higher up in the file than the line on which it is declared

Reply:
Do we keep this lemma then?

Reply:
No, I think remove it

Reply:
Doesn't `eq_of_empty` make users easier to use? (As to prove this, I need first to write `local attribute [instance] matrix.subsingleton_of_empty_left matrix.subsingleton_of_empty_right`)

Reply:
Let's let another reviewer decide. Note that your lemma only works on `0x0` matrices, whereas the instances work on `0xn` and `nx0` matrices

Reply:
Good point. I will rewrite to generalize it.